databroker/assets/core.py

```diff
@@ -9,8 +9,33 @@
 import pandas as pd
 from ..utils import sanitize_np, apply_to_dict_recursively
 
+
 class DatumNotFound(Exception):
-    pass
+    """
+    Raised if a Datum id is not found.
+    """
+    def __init__(self, datum_id, *args):
+        super().__init__(*args)
+        self.datum_id = datum_id
+
+    def __str__(self):
```

Comment:
Overriding `__str__` seems more aggressive than necessary. If one prints this object in a debugging session, it will appear to be a string rather than an Exception object, which would be confusing. Maybe providing a default value for the Exception's `msg` would serve the need here.

```py
In [1]: class Thing(Exception): 
   ...:     def __init__(self, msg=None, *args): 
   ...:         if msg is None: 
   ...:             msg = "My default msg" 
   ...:         super().__init__(msg, *args) 
   ...:                                                                                                                                                                                                                                                                                                                                

In [2]: raise Thing()                                                                                                                                                                         
---------------------------------------------------------------------------
Thing                                     Traceback (most recent call last)
<ipython-input-3-8bca5b9ebda8> in <module>
----> 1 raise Thing()

Thing: My default msg

In [3]: raise Thing("A different msg")                                                                                                                                                        
---------------------------------------------------------------------------
Thing                                     Traceback (most recent call last)
<ipython-input-4-25c6e6d5f226> in <module>
----> 1 raise Thing("A different msg")

Thing: A different msg
```